### ISSUE TITLE
Assertion failed for distance calculator

### ISSUE THREAD

**Issue by spencerwuwu** [aflgo/aflgo#97]
Hello alfgo authors,

I am evaluating aflgo with fuzzing benchmark [Magma](https://github.com/HexHive/magma).
When I tried to use `scripts/gen_distance_fast.py` in building Magma's targets, a lot of computations failed with the assertion
```
distance_calculator: main.cpp:174: std::vector<vertex_desc> cfg_calculation(graph_t &, std::ifstream &, std::ifstream &, std::ifstream &, unordered_map<std::string, double> &, unordered_map<std::string, double> &): Assertion `splits.size() == 2' failed.
```
On the other hand `scripts/genDistance.sh` worked in general, just being very very slow.

May I ask what is this assertion for and what maybe the cause of it? Is it fine to ignore it?

Please let me know if you would like to see the building scripts, binary, etc for this issue.
This happened for both the old aflgo and the current llvm11 version.

Thank you in advance!

**Comment by mboehme** [comment]
Thanks for reporting the issue!
Tagging @Liblor.

**Comment by Liblor** [comment]
Hi @spencerwuwu 

This is probably caused by an empty line in the `callgraph.distance.txt` file (or it is possibly empty), could you have a look at that file and share it? In any case it would also be helpful if you can share your build scripts.


When it comes to your question what it does:
https://github.com/aflgo/aflgo/blob/6e87b69ab2c31177d66e1c3d0b674393cfdceec7/distance_calculator/main.cpp#L158-L164
Basically, the `callgraph.distance.txt` file should have a `basic-block name` and a distance (float) per line separated by a comma, the assertion asserts that exactly two elements are per line. This isn't proper error handling, and was rather used during developing and I'm a bit surprised that it is triggerable with the `gen_distance_fast.py` script.

Best,
Liblor

**Comment by spencerwuwu** [comment]
Hi Liblor and Marcel,
Thank you for the reply!

I checked the `callgraph.distance.txt` in my building environment. It seems to be fine as following:
```
MAGMA_png_read_info,3
MAGMA_png_read_row,4
MAGMA_png_read_end,3
MAGMA_png_read_chunk_header,2
MAGMA_png_check_chunk_length,1
MAGMA_png_read_IDAT_data,3
MAGMA_png_read_finish_IDAT,4
MAGMA_png_read_finish_row,5
LLVMFuzzerTestOneInput,4
```
And, but, somehow, the scripts started to worked today... really curious.
They have been not working for the past few weeks, until I tried to clean it up and send you guys.
I will let you know if I figure out what's happening...

**Comment by Liblor** [comment]
Hi spencerwuwu

Great! In that case, there was probably some polluted state. While this shouldn't happen I didn't do extensive testing in that regard for the `gen_distance_fast.py` script. Happy to look into it, if you figure out what happened or how to reproduce.

**Comment by spencerwuwu** [comment]
Hi Liblor,

I found out that it's only luck that the scripts build a target successfully.
When I run it several times, the assertion is triggered constantly. My guess is that there are some object files that will trigger the assertion, but they may not be processed every time when calculating distances.

I clean up the scripts to demo this in branch `aflgo` of this [repo](https://github.com/spencerwuwu/magma/tree/aflgo).
Everything is built inside a Docker. You can use `cd tools/caption && ./build_aflgo_example.sh` to build the image.
The scripts basically fetch the code of aflgo and libpng into a container's `$FUZZER/repo` and `$TARGET/repo` respectively.
Then, `fuzzers/aflgo/instrument.sh` (`$FUZZER/instrument.sh` in the docker image) is executed to build libpng and calculate distances.

Regardless of the scripts succeed or not, we should get a docker image with or without an image name.
What I did is I run this image, edit `$FUZZER/repo/scripts/gen_distance_fast.py` to print out the `cmd` of `exec_distance_prog`, then run `$FUZZER/instrument.sh`. It's very likely to see `gen_distance_fast.py` failed, with the latest command executed printed.
If the script succeed, well you're lucky lol. Run it again and you should see it fail.

It's a bit complicated to reproduce. Please don't make it a priority to your tasks if you're busy. And thank you in advance again!

Best,
Spencer

**Comment by Liblor** [comment]
Hi Spencer

So today I finally had time to investigate, but I can't reproduce the issue.

Here's what I did:

```bash
git clone https://github.com/spencerwuwu/magma.git aflgo_issue_97
cd aflgo_issue_97/tools/caption
./build_aflgo_example.sh
docker -it ... /bin/bash
```

### Stuff from here is run inside the docker image:

First, I recompiled the `distance_calculator` in debug mode:
```bash
cd $FUZZER/repo/distance_calculator/
rm distance_calculator
cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug  # making sure to absolutely have assertions printed
ninja
```

Next, I edited `exec_distance_prog` as follows:
```python
def exec_distance_prog(dot, targets, out, names, cg_distance=None,
                       cg_callsites=None, py_version=False):
    prog = DIST_BIN if not py_version else DIST_PY
    cmd = [prog,
           "-d", dot,
           "-t", targets,
           "-o", out,
           "-n", names]
    print(" ".join(list(map(str, cmd))))
    if cg_distance is not None and cg_callsites is not None:
        cmd.extend(["-c", cg_distance,
                    "-s", cg_callsites])
    pipe = subprocess.PIPE
    r = subprocess.run(cmd, stdout=pipe, stderr=pipe, check=True)
    print(f"stdout: {r.stdout.decode()}")
    print(f"stderr: {r.stderr.decode()}")
```

Finally, I ran `instrument.sh`:
```bash
./instrument.sh |& test.log
```

I did so multiple times (the last step), but `test.log` never contained the string "Assertion". Let me know, if I followed your instructions properly or whether I missed something crucial!

Cheers,
Liblor

**Comment by waugustus** [comment]
Hi, I met the same problem when I tried to use AFLGO to fuzz `libtiff`. With the debug method in https://github.com/aflgo/aflgo/issues/97#issuecomment-869180985, I found the problem was due to the empty line in `cg_callsites_stream` (`BBnames.txt`), instead of `cg_distance_stream` (`callgraph.distance.txt`). And here is the file,

```
# head -n10 $TMP_DIR/BBcalls.txt 

,TIFFErrorExtR
,TIFFReadDirEntryCheckRangeSbyteSlong8
,TIFFReadDirEntryData
,TIFFYCbCrtoRGB
,__assert_fail
TIFFError
addtiffo.c:105,strcmp
addtiffo.c:113,fprintf
addtiffo.c:126,atoi
```

These empty lines can be removed by adding `|grep -v "^$"` in the cleanup process

```
# Clean up
cat $TMP_DIR/BBnames.txt | grep -v "^$" | rev | cut -d: -f2- | rev | sort | uniq > $TMP_DIR/BBnames2.txt && mv $TMP_DIR/BBnames2.txt $TMP_DIR/BBnames.txt
cat $TMP_DIR/BBcalls.txt |grep -v "^$" | sort | uniq > $TMP_DIR/BBcalls2.txt && mv $TMP_DIR/BBcalls2.txt $TMP_DIR/BBcalls.txt
```

Additionally, several lines in BBcalls.txt may have incorrect number of comma marks, which also caused the assertion failed,

```
# cat $TMP_DIR/BBcalls.txt |grep ".*,.*,"
tif_dirwrite.c:3542,TIFFDataWidthtif_dir.c:60,setByteArray
tiffcp.c:1420,TIFFFileNametiffmedian.c:136,getopt
tiff2pdf.c:3027,TIFFFileNametiff2ps.c:284,getopt
# cat temp_tiffcrop_6247/BBcalls.txt |grep -v ","
tiff2ps.c:2190
```

I do not know why this happens, and it is hard to identify the correct function name automatically, so I corrected this file manually.
